### PR TITLE
PLNSRVCE-1252 : GitHub webhooks to /proxy endpoint

### PR DESCRIPTION
**Pull request by prietyc123**
GitHub webhook post request to `/proxy` endpoint

Signed-off-by: Priti Kumari [pkumari@redhat.com](mailto:pkumari@redhat.com)

### CODE REVIEW

**Comment by prietyc123** [comment]
@adambkaplan Is it something you meant for opening `/proxy` endpoint to be merged in production. Could you please review

File: pkg/server/server_test.go
Comment (adambkaplan) [inline pkg/server/server_test.go]: We really need to test both paths - `/` and `/proxy`. Since we know `/` will be going away, I think it is fine to copy the code into a new method.

**Review by adambkaplan — COMMENTED** [review]
To answer your question, yes, this is generally the right path.

Looking at the code, we might need to add more logic to the tests to make sure the request is sent to the backend as expected.

**Comment by prietyc123** [comment]
> Looking at the code, we might need to add more logic to the tests to make sure the request is sent to the backend as expected.

Sure 👍 
Will add more tests

**Comment by adambkaplan** [comment]
The httptest package has a lot of great utilities for these purposes: https://pkg.go.dev/net/http/httptest#NewServer

File: pkg/backends/server.go
Comment (adambkaplan) [inline pkg/backends/server.go]: If this code is only intended for tests, please rename the file to `server_test.go`. Go has a built in convention where files with the suffix `_test` are only compiled when running `go test` - they are not compiled with `go build` and thus won't make it into production code.

File: pkg/proxy/proxy.go
Comment (adambkaplan) [inline pkg/proxy/proxy.go:102]: Okay for now. We may decide later on that we need labels on this metric to distinguish the endpoints.

File: pkg/proxy/proxy.go
Comment (adambkaplan) [inline pkg/proxy/proxy.go]: A potential bug here - the copied URL will maintain `/proxy` in its path. This needs to be dropped. Otherwise the proxy will forward to `https://<backend-host>/proxy/`. Tests should verify that `/proxy` is dropped from the path.

File: pkg/backends/server.go
Comment (enarha) [reply pkg/backends/server.go]: I also moved those helper methods in a separate package as part of https://github.com/redhat-appstudio/sprayproxy/pull/33. In my case I put them in a package called test. My point is you can't use exported methods (here `NewTestServer`) from a test file, so after renaming the `backend.go` file to `backend_test.go`, my tests started failing with `undefined: test.NewTestServer`. Here is the explanation I found on StackOverflow "Files matching *_test.go are only compiled when testing the package they are part of. If you are testing package A that uses package B, you won't have access to the _test.go code from package B."
Earlier I had another issue "no non-test Go files in <directory>", but that was easy to overcome by adding a doc.go file.
Hope that saves you some time investigating.

File: pkg/server/server_test.go
Comment (enarha) [inline pkg/server/server_test.go]: This feels more like e2e test as it also includes functionality from the proxy package. That's where the `proxied` response comes from. Again in #33  I've put it in a separate e2e package.

File: pkg/server/server_test.go
Comment (enarha) [inline pkg/server/server_test.go]: Duplication of test name `log 200 response`.

File: pkg/backends/server.go
Comment (prietyc123) [reply pkg/backends/server.go]: This file has the function to create the test server. So IF we rename it with `server_test.go` other test files using this function error out with import package. As the content inside this file has some common function, I moved it to single file.
Does moving this file under `/pkg/test` or `pkg/utils` folder would make sense?

Comment (enarha) [reply pkg/backends/server.go]: I prefer top level test directory and that's what I did in my PR. We already discussed using `utils` and similar names in another PR and IIRC the general sentiment was to avoid them.

**Comment by prietyc123** [comment]
> I also moved those helper methods in a separate package as part of https://github.com/redhat-appstudio/sprayproxy/pull/33. In my case I put them in a package called test. My point is you can't use exported methods (here NewTestServer) from a test file, so after renaming the backend.go file to backend_test.go, my tests started failing with undefined: test.NewTestServer. Here is the explanation I found on StackOverflow "Files matching *_test.go are only compiled when testing the package they are part of. If you are testing package A that uses package B, you won't have access to the _test.go code from package B."
Earlier I had another issue "no non-test Go files in ", but that was easy to overcome by adding a doc.go file.
Hope that saves you some time investigating.

Thanks @enarha This was really helpful. I was also searching for how to make it work with `_test.go` file. You really saved my day 🙂

File: pkg/proxy/proxy.go
Comment (prietyc123) [reply pkg/proxy/proxy.go]: Fixed

File: pkg/server/server_test.go
Comment (prietyc123) [reply pkg/server/server_test.go]: Fixed

File: pkg/proxy/proxy_test.go
Comment (adambkaplan) [inline pkg/proxy/proxy_test.go]: +1 on this check here - ensure that we are dropping the `/proxy` path prefix.

File: pkg/proxy/proxy.go
Comment (enarha) [inline pkg/proxy/proxy.go]: If you just want to strip the path from the URL, using the "net/url" makes it more robust and readable. You can either construct the new URL 
`reqURL := "https://localhost:8081/proxy"; u, _ := url.Parse(reqURL); fmt.Sprintf("%s://%s", u.Scheme, u.Host)`
or you can capture the path `u.Path` and strip it from the original URL using strings.TrimSuffix.

File: pkg/server/server.go
Comment (enarha) [inline pkg/server/server.go:65]: Why do we need 3 endpoints handled by the same handler (handleHealthz). That's probably more of a question to @adambkaplan . Also why do we need 2 endpoints, `/` and `/proxy` handled by `sprayProxy.HandleProxy`. I thought `/proxy` is replacing `/`.

File: pkg/proxy/proxy_test.go
Comment (enarha) [inline pkg/proxy/proxy_test.go:83]: This test and the next one quite similar with only difference I see the path in the URL. Maybe combine it in one test and run the 2 cases in parallel using `t.Run`. Then the single name could be TestProxyHandler. I also questioned if we need to handle more than one endpoint here (in the `server.go` file).

File: pkg/server/server_test.go
Comment (enarha) [inline pkg/server/server_test.go]: Almost identical test in the e2e package. I think it rather belongs there.

File: pkg/server/server.go
Comment (prietyc123) [reply pkg/server/server.go:65]: @enarha Once this pr merged we have planned to deprecate the `/` . Refer: [comment](https://github.com/redhat-appstudio/sprayproxy/pull/30#discussion_r1155964614) and [comment](https://github.com/redhat-appstudio/sprayproxy/pull/28#pullrequestreview-1367329680)

File: pkg/proxy/proxy_test.go
Comment (prietyc123) [reply pkg/proxy/proxy_test.go:83]: As we will be deprecating `/` and will be refactored after this pr merge. So not sure if including this in one test still make sense.

File: pkg/server/server_test.go
Comment (prietyc123) [reply pkg/server/server_test.go]: I think in e2e I am verifying the behaviour of the endpoint with response `proxied` and here just the status of the endpoint. However e2e test does verifies this step as well. So I also think removing it from here make sense.

File: pkg/proxy/proxy.go
Comment (prietyc123) [reply pkg/proxy/proxy.go]: Done

File: pkg/server/server_test.go
Comment (enarha) [reply pkg/server/server_test.go]: Yes, that's what I meant (belongs there, to the e2e package).

File: pkg/proxy/proxy.go
Comment (adambkaplan) [inline pkg/proxy/proxy.go]: This will not work - in fact this uncovers gaps in our testing:

- Request headers need to be preserved. These convey important information, especially from GitHub. For example, webhook authentication via the GitHub signature header will break once the request is forwarded to PaC on the member cluster.
- Paths after `/proxy/` need to be preserved.

The better approach is to update `newURL.Path` by using https://pkg.go.dev/strings#TrimPrefix. Here's a quick go playground that demonstrates how I think it can work: https://go.dev/play/p/W38wK--GSDk

**Review by adambkaplan — CHANGES_REQUESTED** [review]

File: pkg/proxy/proxy.go
Comment (prietyc123) [reply pkg/proxy/proxy.go]: > Request headers need to be preserved. These convey important information, especially from GitHub. For example, webhook authentication via the GitHub signature header will break once the request is forwarded to PaC on the member cluster.

ohh yes as we have added the request-ids it will break those. 
Fixed. I have modified the test as well accordingly.

**Comment by prietyc123** [comment]
Some unrelated issue
```code
[2/2] STEP 1/4: FROM registry.access.redhat.com/ubi9-minimal:9.1.0
Trying to pull registry.access.redhat.com/ubi9-minimal:9.1.0...
Error: creating build container: initializing source docker://registry.access.redhat.com/ubi9-minimal:9.1.0: can't talk to a V1 container registry
```


/test sprayproxy-unit-test-on-pull-request

**Comment by prietyc123** [comment]
@adambkaplan Could you please re-review this.

File: pkg/proxy/proxy.go
Comment (adambkaplan) [inline pkg/proxy/proxy.go]: Need to rebase and include #43 logic.

You can also treat this as an opportunity to create a separate function that handles the core proxying logic. This is known as DRYing out your code (Don't Repeat Yourself).

Comment (prietyc123) [reply pkg/proxy/proxy.go]: Done

**Comment by prietyc123** [comment]
@enarha Could you please help reviewing this.

**Comment by prietyc123** [comment]
bump @enarha @adambkaplan for review

File: test/e2e/e2e_server_test.go
Comment (adambkaplan) [inline test/e2e/e2e_server_test.go]: Is there a reason why this test was changed to use the `/proxy` endpoint?

**Review by adambkaplan — COMMENTED** [review]
Generally looks good. I have one question about an e2e test that was changed. My concern is that we lost critical test coverage with the update.

File: test/e2e/e2e_server_test.go
Comment (prietyc123) [reply test/e2e/e2e_server_test.go]: You are right. My views while replacement was to test everything with `/proxy` so that after this change once we deprecate we have testing in place. However I think it makes sense to just add test for `/proxy` separately and let other test do test on `/` itself.
Once we deprecate we will do the required change from test perspective. Does that sounds good @adambkaplan

Comment (adambkaplan) [reply test/e2e/e2e_server_test.go]: makes sense - I think we can lgtm once this change here is reverted.

Comment (prietyc123) [reply test/e2e/e2e_server_test.go]: Fixed

File: pkg/proxy/proxy.go
Comment (enarha) [inline pkg/proxy/proxy.go]: I'll suggest to move that to `HandleProxyEndpoint`. So instead of passing the `nonRootEndpoint` parameter to `handleProxyCommon`, we can simply alter the URL path there. So it'll become:
```
func (p *SprayProxy) HandleProxyEndpoint(c *gin.Context) {
        // comment goes here
        c.Request.URL.Path = strings.TrimPrefix(c.Request.URL.Path, "/proxy")
        handleProxyCommon(p, c)
}
```
That way we keep `handleProxyCommon` "common" and we skip that extra parameter.

Comment (prietyc123) [reply pkg/proxy/proxy.go]: @enarha I have added the fix in [commit](https://github.com/redhat-appstudio/sprayproxy/pull/30/commits/7f2ddc624272ecfe239e51c6a3188b4a44b9a15d) . Let me know if everything looks good I will squash it.

**Comment by prietyc123** [comment]
```
There was an issue validating the commit: "GET https://api.github.com/repos/redhat-appstudio/sprayproxy/collaborators: 403 Resource not accessible by integration []"
```

/retest

File: pkg/proxy/proxy.go
Comment (enarha) [reply pkg/proxy/proxy.go]: Yes. Please squash the commits and I'll lgtm.

Comment (prietyc123) [reply pkg/proxy/proxy.go]: Done

**Pull request merged**
